### PR TITLE
refactor handling of binary artifacts such that they do not depend on naming conventions

### PR DESCRIPTION
**Pull request by tomerd**
motivation: imporve the experience of using binary dependencies

changes:
* track binary artifacts type in workspace state, requires new version (v6) of state file
* derive binary artifact type from content of artifact, not name of directory
* update callsites and tests

radar/93973163

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]
Generally looks great, thanks!

I do wonder about two things:
- we're supporting one level of nesting which seems fine, but I'm not sure if that is entirely obvious (similar to the previous state of requiring nesting which was the reason for this change). Should we collect the directories we looked at and include them in a diagnostic if we didn't end up finding any usable artifacts anywhere so that there's no ambiguity about where we looked?
- in the edge case of someone putting multiple artifact bundles or XCFrameworks into a single ZIP file, we'll match the first one we find, if I understand correctly. Should we instead always look through all directories and have a diagnostic for multiple matches?

File: Sources/Workspace/Workspace+BinaryArtifacts.swift [hunk 1]
```diff
@@ -459,6 +473,56 @@ extension Workspace.BinaryArtifactsManager {
...
+
+    private static func deriveBinaryArtifactKind(fileSystem: FileSystem, path: AbsolutePath) throws -> BinaryTarget.Kind? {
+        let files = try fileSystem.getDirectoryContents(path)
+            .map{ path.appending(component: $0) }
+            .filter { fileSystem.isFile($0) }
+
+        if let infoPlist = files.first(where: { $0.basename.lowercased() == "info.plist" }) {
+            let decoder = PropertyListDecoder()
+            let data: Data = try fileSystem.readFileContents(infoPlist)
+            if let _ = try? decoder.decode(XCFrameworkMetadata.self, from: data) {
```
Comment (neonichu) [inline Sources/Workspace/Workspace+BinaryArtifacts.swift:510]: Should we emit a warning/debug diagnostic for Info.plists we couldn't decode? Could be useful for debugging.

File: Sources/Workspace/Workspace+BinaryArtifacts.swift [hunk 2]
```diff
@@ -459,6 +473,56 @@ extension Workspace.BinaryArtifactsManager {
...
+                return .xcframework
+            }
+        }
+
+        if let infoJSON = files.first(where: { $0.basename.lowercased() == "info.json" }) {
+            if let _ = try? ArtifactsArchiveMetadata.parse(fileSystem: fileSystem, rootPath: infoJSON.parentDirectory) {
```
Comment (neonichu) [inline Sources/Workspace/Workspace+BinaryArtifacts.swift:516]: Same comment as with Info.plist, should we emit the potential error here somewhere?

**Comment by tomerd** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

**Pull request merged**